### PR TITLE
Added fabric-screen-v1 module

### PR DESCRIPTION
**Pull request by Sollace**
Here's the promised new version of that screens module.

Rewrite with some changes
 - Added a dedicated button list that keeps the buttons/elements lists in sync
 - Renamed ScreenView -> ScreenAccess
 - Use the screen itself in the event
 - Provide the client as a useful addition

### CODE REVIEW

**Comment by modmuss50** [comment]
This is looking better than the last one, ill take a better look soon.

File: fabric-screen-v1/src/main/java/net/fabricmc/fabric/api/event/client/screen/ScreenInitCallback.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ * whenever opened. Mods consuming this event are allowed to interact with
+ * the screen to add their own buttons and elements.
+ */
+@FunctionalInterface
+public interface ScreenInitCallback {
+	/**
+	 * Event bus for mods to subscribe to this event.
+	 *
+	 * Usage:
+	 *   ScreenInitCallback.EVENT.subscribe((client, screen) -> {...});
```
Comment (Pyrofab) [inline fabric-screen-v1/src/main/java/net/fabricmc/fabric/api/event/client/screen/ScreenInitCallback.java]: Does this actually count as an event bus ? Either way, you should wrap your code example in `<pre><code>` tags for proper formatting.

File: fabric-screen-v1/src/main/java/net/fabricmc/fabric/impl/event/client/screen/ButtonList.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	public int size() {
+		return buttons.size();
+	}
+
+	private void checkIndex(int index) {
+		if (index < 0) {
+			throw new IndexOutOfBoundsException("index < 0");
+		}
+		if (index > size()) {
+			throw new IndexOutOfBoundsException("index > size()");
```
Comment (Pyrofab) [inline fabric-screen-v1/src/main/java/net/fabricmc/fabric/impl/event/client/screen/ButtonList.java]: These exceptions should indicate the faulty index value to be more useful. They could use the same error message as other lists, like `String.valueOf(index)` for negative indices and `"Index: " + index + ", Size: " + size()` for positive ones.

File: fabric-screen-v1/src/main/java/net/fabricmc/fabric/api/event/client/screen/ScreenInitCallback.java (hunk 1 above)
Comment (Sollace) [reply fabric-screen-v1/src/main/java/net/fabricmc/fabric/api/event/client/screen/ScreenInitCallback.java]: > Does this actually count as an event bus ?

You subscribe to it, and it goes around delivering events to your listener. So yeah, it's a bus.

I'll put in the pre-code tags shortly.

Comment (Chocohead) [reply fabric-screen-v1/src/main/java/net/fabricmc/fabric/api/event/client/screen/ScreenInitCallback.java]: Bus nor not, it's [`Event#register`](https://github.com/FabricMC/fabric/blob/master/fabric-api-base/src/main/java/net/fabricmc/fabric/api/event/Event.java#L51) not `subscribe`

Comment (Sollace) [reply fabric-screen-v1/src/main/java/net/fabricmc/fabric/api/event/client/screen/ScreenInitCallback.java]: Oh whooops. I'll fix that post-haste!

**Comment by Pyrofab** [comment]
Sorry if this is not super constructive but can we get the title fixed please ? It triggers me.

**Comment by valoeghese** [comment]
Yeah. Let's change it to `Addeb fabrig-scren-v1 modrool`

**Comment by Sollace** [comment]
@Pyrofab Oh yeah, thanks for point that out!

**Comment by valoeghese** [comment]
You forgot to change `screen` to `scren`

**Comment by valoeghese** [comment]
Nooooooo

File: fabric-screen-v1/src/main/java/net/fabricmc/fabric/impl/event/client/screen/ButtonList.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	}
+
+	private void rangeCheck(int index) {
+		if (index >= size()) {
+			throw new IndexOutOfBoundsException(outOfBoundsMsg(index));
+		}
+	}
+
+	private void rangeCheckForAdd(int index) {
+		if (index > size() || index < 0) {
```
Comment (InsomniaKitten) [inline fabric-screen-v1/src/main/java/net/fabricmc/fabric/impl/event/client/screen/ButtonList.java]: An index equal to the size is a fail condition too
```suggestion
		if (index >= size() || index < 0) {
```

**Comment by InsomniaKitten** [comment]
`ButtonList` could be using `Preconditions.checkElementIndex` for range preconditioning

File: fabric-screen-v1/src/main/java/net/fabricmc/fabric/impl/event/client/screen/ButtonList.java (hunk 3 above)
Comment (kvverti) [reply fabric-screen-v1/src/main/java/net/fabricmc/fabric/impl/event/client/screen/ButtonList.java]: `add(size(), elem)` is well-defined on Lists.

[Relevant JavaDoc](https://docs.oracle.com/javase/8/docs/api/java/util/List.html#add-int-E-)
> IndexOutOfBoundsException - if the index is out of range (index < 0 || index > size())

**Comment by Sollace** [comment]
@InsomniaKitten + @kvverti 

The checks here are based on the implementations visible in ArrayList and AbstractList. I don't have any intention of deviating from them considering others have specifically asked that I use the same exception messages as the standard list implementations.

**Comment by Pyrofab** [comment]
I mean, I only asked for the message to be more detailed. `Preconditions.checkElementIndex()` seems even better at doing that.

**Comment by Sollace** [comment]
@Pyrofab Ah. Well I added the restriction on top of that. And I'm not in any way inclined to prefer `Preconditions` over just... writing the checks out myself.

It saves a method call and I'd much rather be able to see what's being done than trust some hitherto unknown method. The same reason I don't like `Stream`, or `Optional`, or even `Future`, because all of those replace actual, visible, control flow with a chewed up amalgam of method calls.

**Comment by Player3324** [comment]
Please rebase to 1.15 and make sure it builds (->checkstyle) before we can start reviewing and integrating this.

**Comment by Prospector** [comment]
@Sollace Do you plan on updating this PR or should it be closed?

**Comment by Sollace** [comment]
@Prospector tbh I totally forgot this existed. Is there anything relevant I need to do to update it?

**Comment by Sollace** [comment]
@Sollace 
Me: *posts comment*
Github Notifications: HEY @Sollace , @Sollace JUST COMMENTED ON YOUR PR

smh

**Comment by Prospector** [comment]
Tbh I haven't looked at this PR really I was just going thru old ones and saw Player's last request of a rebase on 1.15 and making sure it passes checkstyle before reviewing continues

**Comment by i509VCB** [comment]
Superseded by #645

**Pull request closed without merging**